benches/cub/device/select/if.cu

```diff
@@ -0,0 +1,176 @@
+#include <nvbench/nvbench.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/sequence.h>
+// Why is this in detail?
+#include <thrust/detail/raw_pointer_cast.h>
+
+#include <cub/device/device_select.cuh>
+
+#include <tbm/range_generator.cuh>
+
+enum class select_op_type
+{
+  greater_than_middle,
+  greater_than_zero,
+  even
+  // complex
+};
+
+template <typename T>
+struct gt_select_op
+{
+  T m_val {};
+
+  explicit gt_select_op(T val)
+      : m_val(val)
+  {}
+
+  __device__ bool operator()(const T& val)
+  {
+    return val > m_val;
+  }
+};
+
+template <typename T>
+struct even_select_op
+{
+  __device__ bool operator()(const T& val)
+  {
+    return (val % 2) == 0;
+  }
+};
+
+template <select_op_type op_type>
+struct op_construction_helper;
+
+template <>
+struct op_construction_helper<select_op_type::greater_than_middle>
+{
+  template <typename T>
+  static gt_select_op<T> create_select_op(int elements)
+  {
+    return gt_select_op<T> (elements / 2);
+  }
+};
+
+template <>
+struct op_construction_helper<select_op_type::greater_than_zero>
+{
+  template <typename T>
+  static gt_select_op<T> create_select_op(int /* elements */)
+  {
+    return gt_select_op<T> (0);
+  }
+};
+
+template <>
+struct op_construction_helper<select_op_type::even>
+{
+  template <typename T>
+  static even_select_op<T> create_select_op(int /* elements */)
+  {
+    return even_select_op<T> ();
+  }
+};
+
+template <typename T, select_op_type SelectOpType, tbm::data_pattern Pattern>
+static void basic(nvbench::state &state,
+                  nvbench::type_list<T,
+                                     nvbench::enum_type<SelectOpType>,
+                                     nvbench::enum_type<Pattern>>)
+{
+  const auto elements = static_cast<int>(state.get_int64("Elements"));
+
+  auto input =
+    tbm::make_range_generator<T, tbm::iterator_style::pointer, Pattern>(
+      elements);
+
+  thrust::device_vector<T> output(elements);
+  thrust::device_vector<T> num_selected(1);
+
+  state.add_element_count(elements);
+  state.add_global_memory_reads(elements);
```

Comment:
Missing a template parameter -- without specifying the type of the reads/writes, NVBench assumes that the argument is the raw number of bytes in the buffer. Also, use `range_generator::get_allocation_size` when possible. Some of the range generators compute values on-the-fly, and thus don't actually contribute to global memory bandwidth.

The number of writes will also depend on how many selected values there are. This should look something like:

```
state.add_element_count(elements);
state.add_global_memory_reads(input.get_allocation_size());
state.add_global_memory_writes<T>(selected_count);
```

Reply:
You can also move this bit after the kernel launch, in which case you can use the `num_selected` result to get an accurate count of the number of writes.